state/protocol/protocol_state/kvstore/statemachine.go

```diff
@@ -0,0 +1,104 @@
+	parentState protocol_state.Reader
+	state       protocol_state.API
+	params      protocol.GlobalParams
+}
+
+var _ protocol_state.KeyValueStoreStateMachine = (*ProcessingStateMachine)(nil)
+
+// NewProcessingStateMachine creates a new key-value store state machine.
+// The underlying state is a clone of the parent state to ensure that the parent state is not modified.
+func NewProcessingStateMachine(
+	view uint64,
+	params protocol.GlobalParams,
+	parentState protocol_state.Reader,
+	mutator protocol_state.API,
+) *ProcessingStateMachine {
+	return &ProcessingStateMachine{
+		view:        view,
+		parentState: parentState,
+		state:       mutator.Clone(),
+		params:      params,
+	}
+}
+
+// Build returns updated key-value store model, state ID and a flag indicating if there were any changes.
+func (m *ProcessingStateMachine) Build() (updatedState protocol_state.Reader, stateID flow.Identifier, hasChanges bool) {
+	updatedState = m.state.Clone()
+	stateID = updatedState.ID()
+	hasChanges = stateID != m.parentState.ID()
+	return
+}
+
+// ProcessUpdate updates the current state of key-value store.
+// KeyValueStoreStateMachine captures only a subset of all service events, those that are relevant for the KV store. All other events are ignored.
+// Implementors MUST ensure KeyValueStoreStateMachine is left in functional state if an invalid service event has been supplied.
+// Expected errors indicating that we have observed and invalid service event from protocol's point of view.
+//   - `protocol.InvalidServiceEventError` - if the service event is invalid for the current protocol state.
+func (m *ProcessingStateMachine) ProcessUpdate(update *flow.ServiceEvent) error {
+	switch update.Type {
+	case flow.ServiceEventProtocolStateVersionUpgrade:
+		versionUpgrade, ok := update.Event.(*flow.ProtocolStateVersionUpgrade)
+		if !ok {
+			return fmt.Errorf("internal invalid type for ProtocolStateVersionUpgrade: %T", update.Event)
+		}
+
+		// To switch the protocol version, replica needs to process a block with a view >= activation view.
+		// But we cannot activate a new version till the block containing the seal is finalized because we cannot switch between chain forks.
+		// The problem is that finality is local to each node due to the nature of the consensus algorithm itself.
+		// We would like to guarantee that all nodes switch the protocol version at exactly the same block.
+		// To guarantee that all nodes switch the protocol version at exactly the same block, we require that the
+		// activation view is higher than the sealing view + Δ when accepting the event. Δ represents the finalization lag
+		// to give time for replicas to finalize the block containing the seal for the version upgrade event.
+		// When replica reaches activation view and the latest finalized protocol state knows about the version upgrade,
+		// then it's safe to switch the protocol version.
+		if m.view+m.params.EpochCommitSafetyThreshold() >= versionUpgrade.ActiveView {
```

Comment:
I agree with using this existing parameter, as the underlying assumption is the same. I suggest we change the name to something like `FinalizationSafetyThreshold`.